### PR TITLE
Unit testing harness for ComponentManager

### PR DESCRIPTION
**Pull request by louiseschmidtgen**
## Description
This PR adds unit test harness(fixtures and mocks) to our `ComponentManager` utility which interacts with the helm client/cli to operate on components. 

#### Interface
In order to be able to override the mocked Helm action configuration, @bschimke95  advised me to use an `HelmConfigInitializer` interface that calls the real or fake `InitializeHelmClientConfig` function. This function either returns a fake action Configuration or the real one. 
The interface is added to the Helm Client and set when the new Component manager instance is created. This required adding the initializer as an argument to the Component manager and update references to the function.

#### Mock Snap
The mock for the snap builds on the snap mock in this repo. I used the snap mock's pathprefix to be able to give the snap a fake `k8s/components/components.yaml` file to build it's configuration from.

#### Mock Action
My Mock for the `helmClient.actionConfig` is heavily inspired by the helm action package and its associated tests and uses parts to create the `actionConfig`, as well as mocks for component releases.

#### Test Set-up
The `createNewManager` function is supposed to make adding new tests easier and is supposed to provide the default HelmClient mock to test the other functions with. I'm open to suggestions on making it easier to use all the mocks and fixtures since there are indeed a lot of them in here.

#### The Bug
During unit testing a little bug crawled out of the List() function:

Instead of setting `allComponents[index] = Component{Name: name}`, I changed it to  `allComponents = append(allComponents, Component{Name: name})` to avoid overwriting components at the same index.

Previously, the `componentsMap` did not change its length so the index didn’t increment so we were just overwriting the component.
```
for name, component := range h.config {
		index := len(componentsMap)
```

[Bug Fix Commit](https://github.com/canonical/k8s-snap/commit/fb5eae50ac7bb415f66300f553eb3172daa1c2fc)

We could consider doing the bug-fix in a separate PR.

### CODE REVIEW

File: src/k8s/pkg/component/component.go [hunk 1]
```diff
@@ -142,28 +170,32 @@ func (h *helmClient) isComponentEnabled(name, namespace string) (bool, error) {
...
 
 	list := action.NewList(actionConfig)
 	releases, err := list.Run()
+
 	if err != nil {
 		return nil, fmt.Errorf("failed to list components: %w", err)
 	}
 
-	allComponents := make([]Component, len(h.config))
+	allComponents := make([]Component, 0)
```
Comment (bschimke95) [inline src/k8s/pkg/component/component.go:185]: A bit of Go nerding:
you can also specify the capacity of the backing array and still set a size of 0 for the slice:
```
allComponents := make([]Component, 0, len(h.config)
```
See: https://goplay.tools/snippet/-pCq4-6OwX8

This is totally crucial for the performance of the k8s-snap and not premature optimization! ;)

File: src/k8s/pkg/component/component_test.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return &action.Configuration{
+		Releases:       storage.Init(driver.NewMemory()),
+		KubeClient:     &kubefake.FailingKubeClient{PrintingKubeClient: kubefake.PrintingKubeClient{Out: io.Discard}},
+		Capabilities:   chartutil.DefaultCapabilities,
+		RegistryClient: registryClient,
+		Log: func(format string, v ...interface{}) {
```
Comment (bschimke95) [inline src/k8s/pkg/component/component_test.go:38]: nit: prefer `any` over interfaces{} according to Go style-guide
Also below

File: src/k8s/pkg/component/component_test.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+  namespace: "default"
+`
+
+func createTemporaryTestDirectory(t *testing.T) string {
+	// Create a temporary test directory to mock the snap
+	// <tempDir>
+	// └── k8s/components
+	// 	├── charts
+	// 	└── component.yaml
+	tempDir, err := os.MkdirTemp("", "test")
```
Comment (bschimke95) [inline src/k8s/pkg/component/component_test.go:135]: `t.TempDir()` is your friend. This will also make sure that tests don't interfere etc.

File: src/k8s/pkg/component/component_test.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+func addConfigToTestDir(t *testing.T, path string, data string) {
+	// Create a file and add some configs
+	err := os.WriteFile(path, []byte(data), 0644)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func createNewManager(t *testing.T, components string) (*helmClient, string, *action.Configuration) {
```
Comment (bschimke95) [inline src/k8s/pkg/component/component_test.go:163]: nit: Go prefers `mustxxx` for setup steps that fatally fail, e.g. `mustCreateNewManager` and `mustCreateTestDir`

File: src/k8s/pkg/component/component_test.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+	for _, rel := range []*release.Release{relStub1, relStub2, relStub3} {
+		if err := store.Create(rel); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	all, err := store.ListReleases()
+	assert.NoError(t, err)
+	assert.Len(t, all, 3, "sanity test: three items added")
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component_test.go:109]: I understand this is using testify, but let's keep `gomega` for all our tests. Have a look at https://github.com/canonical/k8s-snap/blob/main/src/k8s/pkg/k8sd/database/kubernetes_auth_tokens_test.go in order to adjust this

```suggestion
	assert.NoError(t, err)
	assert.Len(t, all, 3, "sanity test: three items added")
```

File: src/k8s/pkg/component/component_test.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+func createTemporaryTestDirectory(t *testing.T) string {
+	// Create a temporary test directory to mock the snap
+	// <tempDir>
+	// └── k8s/components
+	// 	├── charts
+	// 	└── component.yaml
+	tempDir, err := os.MkdirTemp("", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component_test.go:138]: you can use `t.TempDir()`

File: src/k8s/pkg/component/component_test.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+	"helm.sh/helm/v3/pkg/chartutil"
+	kubefake "helm.sh/helm/v3/pkg/kube/fake"
+	"helm.sh/helm/v3/pkg/registry"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/storage"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	"helm.sh/helm/v3/pkg/time"
+)
+
+var verbose = flag.Bool("test.log", false, "enable test logging")
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component_test.go:23]: no need for this, we can just rely on `go test -v` which prints all logs.

File: src/k8s/pkg/component/component_test.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+	// Call the List function with the mock HelmClient
+	components, err := mockHelmClient.List()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.NotNil(t, components)
+	assert.Equal(t, 3, len(components))
+
+	assert.Contains(t, components, Component{Name: "one", Status: true})
```
Comment (bschimke95) [inline src/k8s/pkg/component/component_test.go:239]: Our APIs should be deterministic. Let's sort the output of list alphabetically.

File: src/k8s/pkg/component/component.go [hunk 9]
```diff
@@ -44,52 +54,68 @@ type Component struct {
...
-func NewManager(snap snap.Snap) (*helmClient, error) {
+func NewManager(snap snap.Snap, initializers ...HelmConfigInitializer) (*helmClient, error) {
+	var initializer HelmConfigInitializer
+
+	if len(initializers) > 0 {
+		initializer = initializers[0]
+	} else {
+		// If no initializer provided, use a default one
+		initializer = &HelmClientIntitializer{}
+	}
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component.go:89]: Is this to allow `NewManager(snap)` to just work and not having to update callers? I'd much refer to be explicit here, and update callers as needed. Let's also make this `NewHelmManager`, to indicate that this uses helm underneath and not a generic implementation (the arguments already are helm specific)

```suggestion
func NewHelmManager(snap snap.Snap, initializer HelmConfigInitializer) (*helmClient, error) {
	if initializer == nil {
		initializer = &HelmClientIntitializer{}
	}
```

**Review by bschimke95 — COMMENTED** [review]
Great work @louiseschmidtgen!
I did a first pass with some suggestions.

File: src/k8s/pkg/component/component.go [hunk 10]
```diff
@@ -44,52 +54,68 @@ type Component struct {
...
+
+	if err := viper.ReadInConfig(); err != nil {
 		return nil, err
 	}
+
 	config := make(map[string]componentDefinition)
-	err = viper.Unmarshal(&config)
+	if err := viper.Unmarshal(&config); err != nil {
+		return nil, err
+	}
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component.go:102]: probably unrelated to this PR but: I wonder if we should just drop `viper` completely, then define the `map[string]componentDefinition` in the Go code. Any reason to always parse the same static yaml file? Could we do this?

File: src/k8s/pkg/component/component_test.go [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+	err = os.MkdirAll(k8sComponentsDir, 0777)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	k8sComponentsChartsDir := filepath.Join(k8sComponentsDir, "charts")
+	err = os.MkdirAll(k8sComponentsChartsDir, 0777)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component_test.go:150]: just the second `MkdirAll` will create all previous dirs as well

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Great work! Left some improvement suggestions.

File: src/k8s/pkg/component/component.go (hunk 9 above)
Comment (neoaggelos) [reply src/k8s/pkg/component/component.go:89]: This also needs updating all callers, and they should do `NewHelmManager(snap, nil)`

File: src/k8s/pkg/component/component.go [hunk 12]
```diff
@@ -44,52 +54,68 @@ type Component struct {
...
+	return actionConfig, nil
+}
+
 func logAdapter(format string, v ...any) {
 	logrus.Debugf(format, v...)
 }
 
 // NewManager creates a new Component manager instance.
-func NewManager(snap snap.Snap) (*helmClient, error) {
+func NewManager(snap snap.Snap, initializers ...HelmConfigInitializer) (*helmClient, error) {
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/component/component.go:81]: This is also an option instead of passing `nil` default values, pick whichever you prefer. A simple `nil` arg is probably simpler until we have more configs to consider.

File: src/k8s/pkg/component/component.go (hunk 1 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component.go:185]: Ah, good to know! Learnt something new again :)

File: src/k8s/pkg/component/component_test.go (hunk 5 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component_test.go:109]: Thanks for the example!

File: src/k8s/pkg/component/component_test.go [hunk 13]
```diff
@@ -6,6 +6,7 @@ import (
 	"path/filepath"
 	"testing"
 
+	mocks "github.com/canonical/k8s/pkg/component/mock"
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component_test.go:9]: Any reason to rename from `mock`? 
```suggestion
	"github.com/canonical/k8s/pkg/component/mock"
```

File: src/k8s/pkg/component/component_test.go [hunk 14]
```diff
@@ -180,7 +172,7 @@ func TestNewHelmClientWithValidConfig(t *testing.T) {
 	g.Expect(mockHelmClient).ToNot(BeNil())
 	g.Expect(mockHelmClient).To(BeAssignableToTypeOf(&helmClient{}))
 
-	g.Expect(mockHelmClient.initializer).To(BeAssignableToTypeOf(&MockHelmClientInitializer{}))
+	g.Expect(mockHelmClient.initializer).To(BeAssignableToTypeOf(&mocks.HelmClientInitializer{}))
 	g.Expect(mockHelmClient.snap).To(BeAssignableToTypeOf(&mock.Snap{}))
 	g.Expect(mockHelmClient.config).To(HaveLen(3))
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component_test.go:177]: Are these useful tests? What do we validate here?

File: src/k8s/pkg/component/interface.go [hunk 15]
```diff
@@ -0,0 +1,21 @@
+package component
+
+import "helm.sh/helm/v3/pkg/action"
+
+// HelmComponentManager defines an interface for managing k8s components.
+type HelmComponentManager interface {
```
Comment (neoaggelos) [inline src/k8s/pkg/component/interface.go:6]: Let's keep the interface name generic. Only implementations should describe _how_ they implement the interface.
```suggestion
// ComponentManager defines an interface for managing k8s components.
type ComponentManager interface {
```

File: src/k8s/pkg/component/component.go (hunk 9 above)
Comment (neoaggelos) [reply src/k8s/pkg/component/component.go]: ^ this is not addressed, @louiseschmidtgen

**Review by neoaggelos — COMMENTED** [review]
Nice work! Getting there!

File: src/k8s/pkg/component/component.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component.go]: I'm choosing the `nil` option for now but I will keep the functional pattern in mind if there is more configuration needed later.

File: src/k8s/pkg/component/component.go (hunk 9 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component.go]: You are too quick, I was just updating it :)

File: src/k8s/pkg/component/component_test.go (hunk 13 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component_test.go:9]: I'm also importing the snap mock so go gets confused since both the snap mock and the initializer mock get imported. Do you have a better idea?

File: src/k8s/pkg/component/component_test.go (hunk 14 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component_test.go]: I'll remove this one.

File: src/k8s/pkg/component/component_test.go (hunk 13 above)
Comment (neoaggelos) [reply src/k8s/pkg/component/component_test.go:9]: I would recommend `snapmock` and `componentmock`

Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component_test.go:9]: Good call! Thanks

File: src/k8s/pkg/component/component.go (hunk 10 above)
Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component.go]: Are there any cases we would be updating the `component.yaml`, like a chart version update?

Comment (neoaggelos) [reply src/k8s/pkg/component/component.go]: Not for an installed snap, `component.yaml` is static and immutable. The change requires a new build, therefore it's the same as having it defined in the code

File: src/k8s/pkg/component/component.go [hunk 16]
```diff
@@ -12,17 +13,8 @@ import (
...
-	Enable(name string, values map[string]any) error
-	// List returns a list of enabled components.
-	List() ([]Component, error)
-	// Disable disables a component from the cluster.
-	Disable(name string) error
-	// Refresh updates a k8s component.
-	Refresh(name string) error
-}
+// HelmClientIntitializer implements the HelmConfigInitializer interface
+type HelmClientIntitializer struct{}
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/k8s/pkg/component/component.go:17]: for lack of a better name (and seeing the conflict with the interface name), this could be called `defaultHelmConfigProvider struct`

File: src/k8s/pkg/component/component.go [hunk 17]
```diff
@@ -12,17 +13,8 @@ import (
...
-	Enable(name string, values map[string]any) error
-	// List returns a list of enabled components.
-	List() ([]Component, error)
-	// Disable disables a component from the cluster.
-	Disable(name string) error
-	// Refresh updates a k8s component.
-	Refresh(name string) error
-}
+// HelmClientInitializer implements the HelmConfigInitializer interface
+type HelmClientInitializer struct{}
```
Comment (neoaggelos) [inline src/k8s/pkg/component/component.go:17]: So, this is now great, but still having a `*HelmClientInitializer` be a `HelmConfigInitializer` might be hard to parse and identify issues in the future. Let's instead call them:

(Note also that `provider` is a more common term for this pattern instead of `initializer`)

```
HelmConfigInitializer -> HelmConfigProvider
HelmClientInitializer -> defaultHelmConfigProvider  (it can be private, right?)
```

Comment (louiseschmidtgen) [reply src/k8s/pkg/component/component.go]: Thank you, I've made the change.

**Review by neoaggelos — APPROVED** [review]
LGTM!

**Review by bschimke95 — APPROVED** [review]
LGTM, great work!
